Feat blinking light action

cicoub13 opened the pull request:
### Pull Request check-list

To ensure your Pull Request can be accepted as fast as possible, make sure to review and check all of these items:

- [x] If your changes affects code, did your write the tests?
- [x] Are tests passing? (`npm test` on both front/server)
- [x] Is the linter passing? (`npm run eslint` on both front/server)
- [x] Did you run prettier? (`npm run prettier` on both front/server)
- [x] If you are adding a new features/services, did you run integration comparator? (`npm run compare-translations` on front)
- [x] Did you test this pull request in real life? With real devices? If this development is a big feature or a new service, we recommend that you provide a Docker image to the community ([french forum](https://community.gladysassistant.com/)/[english forum](https://en-community.gladysassistant.com/)) for testing before merging.

NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open.

### Description of change

Please provide a description of the change here. It's always best with screenshots, so don't hesitate to add some!

cicoub13 committed:
feat(action): Add blinking lights

cicoub13 committed:
feat(action): Blink lights frontend style

cicoub13 committed:
fix(action): Fix blinking lights

Commit:
fix: Improve frontend and usability

cicoub13 committed:
fix: Fix backend action

cicoub13 committed:
Add tests

cicoub13 committed:
From flash to blink

cicoub13 committed:
Better blinking time management

cicoub13 committed:
Changing fast blink to 200ms

cicoub13 committed:
Fix front issues

cicoub13 committed:
Fix front style

cicoub13 committed:
Fix front style

cicoub13 committed:
Fix test coverage

Pierre-Gilles commented on server/models/scene.js:
```diff
@@ -60,6 +60,8 @@ const actionSchema = Joi.array().items(
       topic: Joi.string(),
       message: Joi.string().allow(''),
+      blinkingTime: Joi.number(),
+      blinkingSpeed: Joi.string().valid('slow', 'medium', 'fast'),
```
Those should be snake_case :)

cicoub13 replied:
Fixed

Pierre-Gilles commented on server/lib/scene/scene.actions.js:
```diff
@@ -110,6 +110,47 @@ const actionsFunc = {
+          self.device.setValue(device, deviceFeature, newValue);
+          timerId = setTimeout(blink, blinkingInterval);
+        }, blinkingInterval);
+        setTimeout(() => {
```
I think we need to make sure that this action is blocking 

Because right now the logic in scene actions is that each action is blocking and we are going to the next block of action only when all actions in the preceding block are finished

cicoub13 replied:
Fixed and tested with a Telegram message after

cicoub13 committed:
feat(action): Take in account PR review

cicoub13 committed:
fixup

cicoub13 committed:
fixup

cicoub13 committed:
fixup

cicoub13 committed:
fix(action): Fix blinking light action

- Updated sinonjs version to 8.1.1 to benefit from tick.async()
- Moved to inline promise with SetTimeout

cicoub13 committed:
Merge branch 'master' into feat-blinking-light-action

cicoub13 committed:
Merge branch 'master' into feat-blinking-light-action

Pierre-Gilles commented on server/lib/scene/scene.actions.js:
```diff
@@ -110,6 +110,48 @@ const actionsFunc = {
+        for (let i = 0; i < blinkingTime; i += blinkingInterval) {
+          newValue = 1 - newValue;
+          await self.device.setValue(device, deviceFeature, newValue);
+          await new Promise((resolve) => setTimeout(resolve, blinkingInterval));
```
As we use bluebird here, you can use Promise.delay()

http://bluebirdjs.com/docs/api/promise.delay.html

cicoub13 replied:
Fixed

Pierre-Gilles commented on front/src/routes/scene/edit-scene/actions/BlinkLightParams.jsx:
```diff
@@ -0,0 +1,130 @@
+import { Component } from 'preact';
```
It seems there is no default on the "blinking_speed", so when I create an action from scratch: 

![Screenshot 2024-02-26 at 09 16 33](https://github.com/GladysAssistant/Gladys/assets/7365207/c226987c-6ea1-4d48-ae5f-f099328e9f5f)

The select shows "lent" but it's not "lent", it's "empty", and it ends up in "200ms" (super fast)

cicoub13 replied:
Fixed 
https://github.com/GladysAssistant/Gladys/pull/2007/files#diff-95d22c09953ef23db328da87b5b661548fe8958c05e5555874dd7ceed4a0841bR60-R62

Pierre-Gilles reviewed (changes requested):
Thanks for this PR ! It works great, just found a small bug + one small code feedback :)

cicoub13 committed:
feat(action): Take in account PR review

cicoub13 committed:
Merge branch 'master' into feat-blinking-light-action

Pierre-Gilles reviewed (approved):
Good for now me! Thanks for the changes 🙏
